In [85]:
# ignore warnings
import warnings
warnings.filterwarnings("ignore")

import numpy as np

from pydataset import data

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier

import matplotlib.pyplot as plt
import seaborn as sns
from Acquire_MIRZ import get_connection, read_csv
from Prepare_mirz import prep_titanic, my_train_test_split, prep_telco_data
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier





What is your baseline prediction? What is your baseline accuracy? remember: your baseline prediction for a classification problem is predicting the most prevelant class in the training dataset (the mode). When you make those predictions, what is your accuracy? This is your baseline accuracy.

In [34]:
uglytitanic = read_csv('Titanic.csv')

In [35]:
myTitanic = prep_titanic(uglytitanic)

In [36]:
myTitanic.survived.value_counts()

0    424
1    288
Name: survived, dtype: int64

In [37]:
myTitanic

survived  pclass   age  sibsp  parch     fare  alone  is_male  Cherbourg  \
0           0       3  22.0      1      0   7.2500      0        1        0.0   
1           1       1  38.0      1      0  71.2833      0        0        1.0   
2           1       3  26.0      0      0   7.9250      1        0        0.0   
3           1       1  35.0      1      0  53.1000      0        0        0.0   
4           0       3  35.0      0      0   8.0500      1        1        0.0   
..        ...     ...   ...    ...    ...      ...    ...      ...        ...   
885         0       3  39.0      0      5  29.1250      0        0        0.0   
886         0       2  27.0      0      0  13.0000      1        1        0.0   
887         1       1  19.0      0      0  30.0000      1        0        0.0   
889         1       1  26.0      0      0  30.0000      1        1        1.0   
890         0       3  32.0      0      0   7.7500      1        1        0.0   

     Southampton  Queenstown  
0            1.0         0.0  
1            0.0         0.0  
2            1.0         0.0  
3            1.0         0.0  
4            1.0         0.0  
..           ...         ...  
885          0.0         1.0  
886          1.0         0.0  
887          1.0         0.0  
889          0.0         0.0  
890          0.0         1.0  

[712 rows x 11 columns]

In [38]:
myTitanic.survived.value_counts()

0    424
1    288
Name: survived, dtype: int64

In [39]:
baseline = len(myTitanic[myTitanic.survived == 0]) / len(myTitanic)
baseline

0.5955056179775281

Fit the decision tree classifier to your training sample and transform (i.e. make predictions on the training sample

In [40]:
train,validate,test = my_train_test_split(myTitanic,'survived')
x_train = train.drop(columns = 'survived')
y_train = train.survived
x_validate = validate.drop(columns = 'survived')
y_validate = validate.survived

In [41]:
DTC = DecisionTreeClassifier(max_depth = 5)
DTC = DTC.fit(x_train,y_train)


In [42]:
y_predictions = DTC.predict(x_train)

Evaluate your in-sample results using the model score, confusion matrix, and classification report.

In [43]:
DTC.score(x_train,y_train)

0.8544600938967136

In [44]:
print(classification_report(y_train,y_predictions))

              precision    recall  f1-score   support

           0       0.83      0.95      0.89       254
           1       0.91      0.71      0.80       172

    accuracy                           0.85       426
   macro avg       0.87      0.83      0.84       426
weighted avg       0.86      0.85      0.85       426



In [45]:
labels = sorted(y_train.unique())
pd.DataFrame(confusion_matrix(y_train, y_predictions), index=labels, columns=labels)

0    1
0  242   12
1   50  122

Compute: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [102]:
Accuracy = (242 + 122) / (242 + 122 + 12 + 50)
TPR = 122 /(122 + 12)
TNR = 242 / (50 + 242)
FPR = 50/ (50 + 242)
FNR = 12 / (12 + 122)
precision = 122 / (122 + 50)
recall = 122 / (122 + 12)
f1 = 122/ 122 + (.5 * (50 + 12))
support_S = 122
support_D = 242

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.8544600938967136 TPR =  0.9104477611940298 TNR =  0.8287671232876712 FPR = 0.17123287671232876 FNR =  0.08955223880597014 Precision =  0.7093023255813954 Recall =  0.9104477611940298 F1 =  32.0 Survived =  122 Died =  242


Run through steps 2-4 using a different max_depth value.

In [47]:
DTC2 = DecisionTreeClassifier(max_depth = 8)

In [48]:
DTC2.fit(x_train,y_train)

DecisionTreeClassifier(max_depth=8)

In [49]:
y_predictions2 = DTC2.predict(x_train)
y_predictions2

array([0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0,

In [50]:
DTC2.score(x_train,y_train)

0.9295774647887324

In [51]:
print(classification_report(y_train,y_predictions2))

              precision    recall  f1-score   support

           0       0.90      0.99      0.94       254
           1       0.98      0.84      0.91       172

    accuracy                           0.93       426
   macro avg       0.94      0.92      0.92       426
weighted avg       0.93      0.93      0.93       426



In [52]:
labels = sorted(y_train.unique())
pd.DataFrame(confusion_matrix(y_train, y_predictions2))

0    1
0  251    3
1   27  145

In [101]:
Accuracy = (251 + 145) / (251 + 145 + 3 + 27)
TPR = 145 /(145 + 3)
TNR = 251 / (27 + 251)
FPR = 27/ (27 + 251)
FNR = 3 / (3 + 145)
precision = 145 / (145 + 27)
recall = 145 / (145 + 3)
f1 = 145/ 145 + (.5 * (27 + 3))
support_S = 145
support_D = 251

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.9295774647887324 TPR =  0.9797297297297297 TNR =  0.9028776978417267 FPR = 0.09712230215827339 FNR =  0.02027027027027027 Precision =  0.8430232558139535 Recall =  0.9797297297297297 F1 =  16.0 Survived =  145 Died =  251


2nd model works better on in-Sample-data


In [54]:
DTC.score(x_validate,y_validate)

0.7692307692307693

In [55]:
DTC2.score(x_validate,y_validate)

0.7622377622377622

first model works better on out of sample data.

In [56]:
uglytelco = read_csv('Telco.csv')

In [57]:
telco = prep_telco_data(uglytelco)

In [58]:
telco = telco.dropna()

In [59]:
telco.did_churn.value_counts()

0    3249
1    1586
Name: did_churn, dtype: int64

In [60]:
baseline = 3249/ len(telco)

In [61]:
baseline

0.6719751809720786

In [62]:
train,validate,test = my_train_test_split(telco,'did_churn')
x_train = train.drop(columns = 'did_churn')
y_train = train.did_churn
x_validate = validate.drop(columns = 'did_churn')
y_validate = validate.did_churn

In [63]:
DTC = DecisionTreeClassifier(max_depth = 5)
DTC = DTC.fit(x_train,y_train)


ValueError: could not convert string to float: ''

In [ ]:
telco

# Random Forest

Fit the Random Forest classifier to your training sample and transform (i.e. make predictions on the training sample) setting the random_state accordingly and setting min_samples_leaf = 1 and max_depth = 10.

In [64]:
train,validate,test = my_train_test_split(myTitanic,'survived')
x_train = train.drop(columns = 'survived')
y_train = train.survived
x_validate = validate.drop(columns = 'survived')
y_validate = validate.survived

In [67]:
RF = RandomForestClassifier(min_samples_leaf = 1, max_depth = 10)

In [68]:
RF.fit(x_train,y_train)

RandomForestClassifier(max_depth=10)

In [69]:
preds = RF.predict(x_train)

Evaluate your results using the model score, confusion matrix, and classification report.

In [72]:
RF.score(x_train,y_train)

0.9694835680751174

In [73]:
print(classification_report(y_train,preds))

              precision    recall  f1-score   support

           0       0.95      1.00      0.98       254
           1       1.00      0.92      0.96       172

    accuracy                           0.97       426
   macro avg       0.98      0.96      0.97       426
weighted avg       0.97      0.97      0.97       426



In [75]:
pd.DataFrame(confusion_matrix(y_train, preds))

0    1
0  254    0
1   13  159

Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [100]:
Accuracy = (254 + 159) / (254 + 159 + 0 + 13)
TPR = 159 /(159 + 0)
TNR = 254 / (13 + 254)
FPR = 13/ (13 + 254)
FNR = 0 / (0 + 159)
precision = 159 / (159 + 13)
recall = 159 / (159 + 0)
f1 = 159/ 159 + (.5 * (13 + 0))
support_S = 159
support_D = 254

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.9694835680751174 TPR =  1.0 TNR =  0.951310861423221 FPR = 0.04868913857677903 FNR =  0.0 Precision =  0.9244186046511628 Recall =  1.0 F1 =  7.5 Survived =  159 Died =  254


In [82]:
x = 20
for i in range (1,21,2):
    RF = RandomForestClassifier(min_samples_leaf = i, max_depth = x )
    RF.fit(x_train,y_train)
    preds = RF.predict(x_train)
    print('min_sample = ',i,'max_depth = ', x, classification_report(y_train,preds))
    x = x - 2

min_sample =  1 max_depth =  20               precision    recall  f1-score   support

           0       0.98      1.00      0.99       254
           1       1.00      0.98      0.99       172

    accuracy                           0.99       426
   macro avg       0.99      0.99      0.99       426
weighted avg       0.99      0.99      0.99       426

min_sample =  3 max_depth =  18               precision    recall  f1-score   support

           0       0.87      0.95      0.91       254
           1       0.91      0.80      0.85       172

    accuracy                           0.89       426
   macro avg       0.89      0.87      0.88       426
weighted avg       0.89      0.89      0.89       426

min_sample =  5 max_depth =  16               precision    recall  f1-score   support

           0       0.86      0.94      0.90       254
           1       0.89      0.78      0.83       172

    accuracy                           0.87       426
   macro avg       0.88      0.8

Best in-sample performance from min_sample_leaf = 1 max_depth = 20

In [84]:
x = 20
for i in range (1,21,2):
    RF = RandomForestClassifier(min_samples_leaf = i, max_depth = x )
    RF.fit(x_train,y_train)
    preds = RF.predict(x_validate)
    print('min_sample = ',i,'max_depth = ', x, classification_report(y_validate,preds))
    x = x - 2

min_sample =  1 max_depth =  20               precision    recall  f1-score   support

           0       0.80      0.85      0.82        85
           1       0.75      0.69      0.72        58

    accuracy                           0.78       143
   macro avg       0.78      0.77      0.77       143
weighted avg       0.78      0.78      0.78       143

min_sample =  3 max_depth =  18               precision    recall  f1-score   support

           0       0.79      0.89      0.84        85
           1       0.81      0.66      0.72        58

    accuracy                           0.80       143
   macro avg       0.80      0.77      0.78       143
weighted avg       0.80      0.80      0.79       143

min_sample =  5 max_depth =  16               precision    recall  f1-score   support

           0       0.76      0.92      0.83        85
           1       0.82      0.57      0.67        58

    accuracy                           0.78       143
   macro avg       0.79      0.7

Best performing out of sample date is at min_sample of 3 and max_depth of 18

# K-Nearest-Neighbors:

Fit a K-Nearest Neighbors classifier to your training sample and transform (i.e. make predictions on the training sample)



In [86]:
KNN = KNeighborsClassifier(n_neighbors = 3)

In [87]:
KNN.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=3)

In [88]:
preds = KNN.predict(x_train)

Evaluate your results using the model score, confusion matrix, and classification report.

In [90]:
KNN.score(x_train,y_train)

0.8450704225352113

In [92]:
print(classification_report(y_train,preds))

              precision    recall  f1-score   support

           0       0.85      0.90      0.87       254
           1       0.84      0.76      0.80       172

    accuracy                           0.85       426
   macro avg       0.84      0.83      0.84       426
weighted avg       0.84      0.85      0.84       426



In [93]:
pd.DataFrame(confusion_matrix(y_train,preds))

0    1
0  229   25
1   41  131

Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [99]:
Accuracy = (229 + 131) / (229 + 131 + 25 + 41)
TPR = 131 /(131 + 25)
TNR = 229 / (41 + 229)
FPR = 41/ (41 + 229)
FNR = 25 / (25 + 131)
precision = 131 / (131 + 41)
recall = 131 / (131 + 25)
f1 = 131/ 131 + (.5 * (41 + 25))
support_S = 131
support_D = 229

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.8450704225352113 TPR =  0.8397435897435898 TNR =  0.8481481481481481 FPR = 0.15185185185185185 FNR =  0.16025641025641027 Precision =  0.7616279069767442 Recall =  0.8397435897435898 F1 =  34.0 Survived =  131 Died =  229


Run through steps 2-4 setting k to 10

10 neighbors:

In [97]:
KNN = KNeighborsClassifier(n_neighbors = 10)
KNN.fit(x_train,y_train)
preds = KNN.predict(x_train)
print(KNN.score(x_train,y_train),  classification_report(y_train,preds))
pd.DataFrame(confusion_matrix(y_train,preds))

0.7582159624413145               precision    recall  f1-score   support

           0       0.75      0.89      0.81       254
           1       0.78      0.56      0.65       172

    accuracy                           0.76       426
   macro avg       0.76      0.73      0.73       426
weighted avg       0.76      0.76      0.75       426



0   1
0  226  28
1   75  97

In [103]:
Accuracy = (226 + 97) / (226 + 97 + 28 + 75)
TPR = 97 /(97 + 28)
TNR = 226 / (75 + 226)
FPR = 75/ (75 + 226)
FNR = 28 / (28 + 97)
precision = 97 / (97 + 75)
recall = 97 / (97 + 28)
f1 = 97/ 97 + (.5 * (75 + 28))
support_S = 97
support_D = 226

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.7582159624413145 TPR =  0.776 TNR =  0.7508305647840532 FPR = 0.24916943521594684 FNR =  0.224 Precision =  0.563953488372093 Recall =  0.776 F1 =  52.5 Survived =  97 Died =  226


20 neighbors:

In [98]:
KNN = KNeighborsClassifier(n_neighbors = 20)
KNN.fit(x_train,y_train)
preds = KNN.predict(x_train)
print(KNN.score(x_train,y_train),  classification_report(y_train,preds))
pd.DataFrame(confusion_matrix(y_train,preds))

0.7370892018779343               precision    recall  f1-score   support

           0       0.74      0.85      0.79       254
           1       0.72      0.56      0.63       172

    accuracy                           0.74       426
   macro avg       0.73      0.71      0.71       426
weighted avg       0.74      0.74      0.73       426



0   1
0  217  37
1   75  97

In [104]:
Accuracy = (217 + 97) / (217 + 97 + 37 + 75)
TPR = 97 /(97 + 37)
TNR = 217 / (75 + 217)
FPR = 75/ (75 + 217)
FNR = 37 / (37 + 97)
precision = 97 / (97 + 75)
recall = 97 / (97 + 37)
f1 = 97/ 97 + (.5 * (75 + 37))
support_S = 97
support_D = 217

print('Accuracy = ', Accuracy , 'TPR = ', TPR , 'TNR = ', TNR , 'FPR =', FPR , 'FNR = ', FNR ,'Precision = ' , precision , 'Recall = ' , recall , 'F1 = ',  f1, 'Survived = ' , support_S , 'Died = ' , support_D )

Accuracy =  0.7370892018779343 TPR =  0.7238805970149254 TNR =  0.7431506849315068 FPR = 0.2568493150684932 FNR =  0.27611940298507465 Precision =  0.563953488372093 Recall =  0.7238805970149254 F1 =  57.0 Survived =  97 Died =  217


What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

My first model with 5 nearest neighbors performed the best on in-sample data

What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

In [106]:
values = [5,10,20]

In [108]:
for x in values:
    KNN = KNeighborsClassifier(n_neighbors = x)
    KNN.fit(x_train,y_train)
    preds = KNN.predict(x_validate)
    print('Num_neighbors: ', x , classification_report(y_validate,preds))

Num_neighbors:  5               precision    recall  f1-score   support

           0       0.68      0.69      0.69        85
           1       0.54      0.52      0.53        58

    accuracy                           0.62       143
   macro avg       0.61      0.61      0.61       143
weighted avg       0.62      0.62      0.62       143

Num_neighbors:  10               precision    recall  f1-score   support

           0       0.67      0.82      0.74        85
           1       0.62      0.41      0.49        58

    accuracy                           0.66       143
   macro avg       0.64      0.62      0.62       143
weighted avg       0.65      0.66      0.64       143

Num_neighbors:  20               precision    recall  f1-score   support

           0       0.68      0.81      0.74        85
           1       0.61      0.43      0.51        58

    accuracy                           0.66       143
   macro avg       0.64      0.62      0.62       143
weighted avg      

My models with 10 and 20 neighbors both performed best at 66% accurate